pkg/cluster/storage.go

```diff
@@ -56,18 +79,108 @@ func (c *Cluster) storageBootstrap(ctx context.Context) error {
 	if err != nil {
 		return err
 	}
 
 	return bootstrap.Read(bytes.NewBuffer(data), &c.runtime.ControlRuntimeBootstrap)
 }
+
+// getBootstrapKeyFromStorage will list all keys that has prefix /bootstrap and will check for key that is
+// hashed with empty string and will check for any key that is hashed by different token than the one
+// passed to it, it will return error if it finds a key that is hashed with different token and will return
+// value if it finds the key hashed by passed token or empty string
+func (c *Cluster) getBootstrapKeyFromStorage(ctx context.Context, storageClient client.Client, token string) (*client.Value, bool, error) {
+	emptyStringKey := storageKey("")
+	tokenKey := storageKey(token)
+
+	bootstrapList, err := storageClient.List(ctx, "/bootstrap", 0)
+	if err != nil {
+		return nil, false, err
+	}
+	if len(bootstrapList) == 0 {
+		c.saveBootstrap = true
+		return nil, false, nil
+	}
+	if len(bootstrapList) > 1 {
+		return nil, false, errors.New("Found more than one bootstrap keys in storage")
+	}
+	bootstrapKV := bootstrapList[0]
+	// checking for empty string bootstrap key
+	if string(bootstrapKV.Key) == emptyStringKey {
+		logrus.Warnf("Bootstrap data already found and encrypted with empty string, deleting empty key")
+		c.saveBootstrap = true
+		if err := storageClient.Delete(ctx, emptyStringKey); err != nil {
+			return nil, false, err
+		}
+		return &bootstrapKV, true, nil
+	} else if string(bootstrapKV.Key) == tokenKey {
+		return &bootstrapKV, false, nil
+	}
+
+	return nil, false, errors.New("Bootstrap data already found and encrypted with different token")
+}
+
+// readTokenFromFile will attempt to get the token from <data-dir>/token if it the file not found
+// in case of fresh installation it will try to use the runtime serverToken saved in memory
+// after stripping it from any additional information like the username or cahash, if the file
+// found then it will still strip the token from any additional info
+func readTokenFromFile(serverToken, dataDir string) (string, error) {
+	tokenFile := filepath.Join(dataDir, "token")
+	b, err := ioutil.ReadFile(tokenFile)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return serverToken, nil
+		}
+		return "", err
+	}
+	return string(b), nil
+}
+
+// normalizeToken will normalize the token read from file or passed as a cli flag
+func normalizeToken(token string) (string, error) {
+	var normalizedToken string
+	// strip the token from any new line if its read from file
+	t := strings.TrimSuffix(string(token), "\n")
+	info, err := clientaccess.ParseToken(t)
+	if err != nil {
+		return normalizedToken, err
+	}
+	// handle runtime.ServerToken being in the format of "server:<token>"
+	// if its in that format then user the <token> only
+	normalizedToken = info.Password
+	parts := strings.SplitN(info.Password, ":", 2)
+	if len(parts) > 1 && parts[0] == "server" {
+		normalizedToken = parts[1]
+	}
+	return normalizedToken, nil
```

Comment:
Are you sure this is necessary? ParseToken already does this splitting for you; info.Password will be just the password, and "server" will go into info.Username.

Also, normalizeToken doesn't seem like the correct name for this function, since what it's really doing is getting just the password from a token string. Maybe call it getPassword but honestly if that's all you want you there's already a function for that:

```go
_, password, ok := clientaccess.ParseUsernamePassword(tokenString)
```

Reply:
yeah but there is a problem when the token comes from c.runtime.ServerToken which is in this format "server:<token>" where parseToken doesnt split those, I didnt want to modify the logic of parsetoken to avoid any possible regression

Reply:
Oh I see, ParseToken expects to be handed in either a full token with the K10 prefix, or just the password. If you pass it in user:password it mangles it a bit.

If you're getting the token from c.runtime.ServerToken you can get a properly formatted token string by calling:
```go
token, err := server.FormatToken(c.runtime.ServerToken,  c.runtime.ServerCA)
if err != nil {
    return err
}
_, password, ok := clientaccess.ParseUsernamePassword(token)
```

Reply:
yes the only problem is that will cause in import cycle if I called formatToken from server, since server calls this package

Reply:
Maybe move FormatToken into the clientaccess package then? It doesn't feel like something that properly belongs in server.

Reply:
cool

Reply:
How many places is it used?

Reply:
just one other place, when grabbing the token to configure the server's local agent:
https://github.com/k3s-io/k3s/blob/master/pkg/cli/server/server.go#L398